### PR TITLE
Multichain

### PR DESCRIPTION
**Pull request by Sergei2404**

### CODE REVIEW

File: near/contract/src/token_trasnferer.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        msg: String,
+    ) -> PromiseOrValue<U128> {
+        self.stake_configuration.assert_token_valid();
+
+        self.activate_reporter(sender_id, amount);
+        PromiseOrValue::Value(U128(0))
+    }
+}
+
+// #[near_bindgen]
```
Comment (hlgltvnnk) [inline near/contract/src/token_trasnferer.rs:47]: Is this part of the code is should be commented? It is better to remove unnecessary comments

File: near/contract/src/reporter/management.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        id: ReporterId,
+        account_id: AccountId,
+        name: String,
+        role: Role,
+        url: String,
+    ) {
+        self.assert_authority();
+
+        require!(self.reporters.get(&id).is_none(), ERROR_REPORTER_EXISTS);
+        require!(
```
Comment (ars9) [inline near/contract/src/reporter/management.rs:26]: Maybe this check is redundant? Or is this an insurance that we indeed added reporter both to the list and the map?

File: near/contract/src/reporter/management.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        url: String,
+    ) {
+        self.assert_authority();
+
+        let mut reporter: Reporter = self
+            .reporters
+            .get(&id)
+            .expect(ERROR_REPORTER_NOT_FOUND)
+            .into();
+        reporter.account_id = account_id;
```
Comment (ars9) [inline near/contract/src/reporter/management.rs:61]: If we change the account ID, we must also modify the reporters_by_account mapping.

File: near/contract/src/stake/mod.rs [hunk 4]
```diff
@@ -42,3 +45,33 @@ impl Default for StakeConfiguration {
...
+    // check if the stake amount is enough for the reporter type
+    pub fn assert_stake_sufficient(&self, amount: U128, role: &Role) {
+        let stake = match role {
+            Role::Validator => self.stake_amounts.validator.0,
+            Role::Tracer => self.stake_amounts.tracer.0,
+            Role::Publisher => self.stake_amounts.publisher.0,
+            Role::Authority => self.stake_amounts.authority.0,
+            Role::Appraiser => 0,
+        };
+        require!(amount.0 >= stake, ERROR_INVALID_STAKE_AMOUNT)
```
Comment (ars9) [inline near/contract/src/stake/mod.rs:64]: The amount must be equal, not just greater.

File: near/contract/src/token_trasnferer.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
+use near_contract_standards::fungible_token::{core::ext_ft_core, receiver::FungibleTokenReceiver};
```
Comment (ars9) [inline near/contract/src/token_trasnferer.rs:1]: Typo in the file name (`token_trasnferer`).

**Review by ars9 — CHANGES_REQUESTED** [review]

File: near/contract/src/reporter/management.rs (hunk 3 above)
Comment (Sergei2404) [reply near/contract/src/reporter/management.rs:61]: Good catch

File: near/contract/src/reporter/management.rs (hunk 2 above)
Comment (Sergei2404) [reply near/contract/src/reporter/management.rs:26]: This check is an insurance that we don't store one account_id for many users_id

**Pull request merged**